### PR TITLE
Remove namespace

### PR DESCRIPTION
**Pull request by Seb-figment**
When creating assemblyScript proto binding from third party proto file that span several file for the same proto package the generated file couldn't be use "AS IS" easily or without modification. 

- The import need to be corrected.
- Files defining message in the same package will all export the same top level namespace. This force the use of "import as".
- We also need to correct the field name when this happen.
- The name conflict and necessity to "import as" is also seen by the user of the produce code.

Here is an [example repo ](https://github.com/Seb-figment/as-proto-example/compare/namespace...no-namespace?diff=split) to show the concern and difference.

example use there: https://github.com/graphprotocol/cosmoshub-ts

### CODE REVIEW

**Comment by Seb-figment** [comment]
@piotr-oles Sorry about all the pending workflow. I'm done pushing on the PR until review comment.
@roaminroe Another approach on what you did in your fork. Didn't knew it existed before I started this.

**Comment by piotr-oles** [comment]
Hi! Thanks for taking time on this PR.
It has a lot of changes and some seems unrelated. Could you try to minimize the diff? And could you add more context on PR motivation?

**Comment by Seb-figment** [comment]
Hi @piotr-oles. The only thing not related to namespace removal is the option to generate dependencies. I have updated the PR description with a link to a diff of change in example repo. I'll leave comment in the PR to explain the modification.

File: packages/as-proto-gen/src/file-context.ts [hunk 1]
```diff
@@ -26,13 +27,19 @@ export class FileContext {
   }
 
   registerImport(importNamePath: string, importPath: string): string {
+    [importNamePath, importPath] = this.getRelativeImportPath(
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/file-context.ts:30]: Make import relative to the current file.

File: packages/as-proto-gen/src/file-context.ts [hunk 2]
```diff
@@ -44,8 +51,57 @@ export class FileContext {
     return [uniqueImportName, ...importNamespace].join(".");
   }
 
-  registerDefinition(definitionNamePath: string): string {
-    const [definitionName] = definitionNamePath.split(".");
+  getRelativeImportPath(
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/file-context.ts:55]: Make import relative to the current file

File: packages/as-proto-gen/src/generate/enum.ts [hunk 3]
```diff
@@ -7,16 +7,17 @@ import * as assert from "assert";
...
   enumDescriptor: EnumDescriptorProto,
-  fileContext: FileContext
+  fileContext: FileContext,
+  enumNamespace?: string
 ): string {
   const enumName = enumDescriptor.getName();
   assert.ok(enumName !== undefined);
 
-  const Enum = fileContext.registerDefinition(enumName);
+  const Enum = fileContext.registerDefinition(enumName, enumNamespace);
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/generate/enum.ts:16]: To handle namespace for enum embeded in a message.

File: packages/as-proto-gen/src/generate/field.ts [hunk 4]
```diff
@@ -18,7 +18,7 @@ export function generateFieldEncodeInstruction(
 
   const fieldTag = getFieldTag(fieldDescriptor);
   const fieldName = generateFieldName(fieldDescriptor);
-  const fieldVariable = scopeContext.getFreeName(fieldName);
+  const fieldVariable = scopeContext.getFreeName(fieldName) + "_";
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/generate/field.ts]: Prevent name collision

File: packages/as-proto-gen/src/generate/file.ts [hunk 5]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+    console.error(error);
+  }
+
+  const outputFile = new CodeGeneratorResponse.File();
+  outputFile.setName(filename);
+  outputFile.setContent(formattedCode);
+  codeGenResponse.addFile(outputFile);
+}
+
+export function generateExport(
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/generate/file.ts:65]: generate all the _export and index to follow message package

File: packages/as-proto-gen/src/index.ts [hunk 6]
```diff
@@ -1,60 +1,93 @@
...
 import { GeneratorContext } from "./generator-context";
-import { generateFile } from "./generate/file";
 import { getPathWithoutProto } from "./names";
-import { FileContext } from "./file-context";
-import prettier from "prettier";
-import * as fs from "fs";
-import * as assert from "assert";
 
-const input = fs.readFileSync(process.stdin.fd);
+fs.readFile(process.stdin.fd, (err, input) => {
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/index.ts:13]: readFileSync was throwing error often on large file. Less frequent this way but still occurs

File: packages/as-proto-gen/src/index.ts [hunk 7]
```diff
@@ -1,60 +1,93 @@
...
-      new FileContext(generatorContext, fileDescriptor)
-    );
-    let formattedCode = generatedCode;
-    try {
-      formattedCode = prettier.format(generatedCode, {
-        parser: "typescript",
-      });
-    } catch (error) {
-      console.error(error);
+    const parameters: Set<string> = new Set(codeGenRequest.getParameter()?.split(','));
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/index.ts]: Automatically generate code for dependencies if `--opt_as=gen-dependencies` is pass argument is pass to protoc.

File: packages/as-proto-gen/src/generator-context.ts [hunk 8]
```diff
@@ -94,6 +95,14 @@ export class GeneratorContext {
     this.typeNameToFileDescriptor.set(fieldTypeName, fileDescriptor);
   }
 
+  addProtoDependencies(fileName: string) {
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/generator-context.ts:98]: To handle generation of code for dependencies

File: packages/as-proto-gen/src/generate/ref.ts [hunk 9]
```diff
@@ -19,11 +20,17 @@ export function generateRef(
   const typeName = getTypeName(fieldTypeName);
 
   if (isSameFile) {
-    return fileContext.registerDefinition(typeName);
+    const types = typeName.split(".");
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/generate/ref.ts:23]: To handle enum/message embeded in a message. (namespaced ref)

File: packages/as-proto-gen/src/generate/file.ts [hunk 10]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
-      const packagePart = packageParts.pop()!; // type assertion - see line above
-      NamespacedTypes = `
-        export namespace ${packagePart} {
-          ${NamespacedTypes}
-        }
-      `;
+  return `${fileContext.getImportsCode()}\n\n${types.join("\n\n")}`;
+}
+
+export function addFile(
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/generate/file.ts:36]: moved from index.ts for reuse

File: packages/as-proto-gen/src/generate/message.ts [hunk 11]
```diff
@@ -71,8 +62,28 @@ export function generateMessage(
...
+      ${generateEncodeMethod(messageDescriptor, fileContext)}
+      ${generateDecodeMethod(messageDescriptor, fileContext)}
+
+      ${generateMessageFieldsDeclarations(messageDescriptor, fileContext)}
+
+      ${generateMessageConstructor(messageDescriptor, fileContext)}
+    }
+  `;
+
+  fileContext.registerImport("Protobuf", "as-proto");
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/generate/message.ts:77]: Adding a utility function to decode message.

File: packages/as-proto-gen/src/generate/message.ts [hunk 12]
```diff
@@ -35,19 +36,7 @@ export function generateMessage(
...
-    export class ${Message} {      
-      ${generateEncodeMethod(messageDescriptor, fileContext)}
-      ${generateDecodeMethod(messageDescriptor, fileContext)}
-      
-      ${generateMessageFieldsDeclarations(messageDescriptor, fileContext)}
-      
-      ${generateMessageConstructor(messageDescriptor, fileContext)}
-    }
-  `;
+  const Message = fileContext.registerDefinition(messageName, messageNamespace);
```
Comment (Seb-figment) [inline packages/as-proto-gen/src/generate/message.ts:39]: new argument to handle embeded message.

**Comment by piotr-oles** [comment]
Thanks for more details and comments! I will take a look hopefully in a few days (I'm a little bit busy now)

File: packages/as-proto-gen/package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "as-proto-gen",
-  "version": "0.2.4",
+  "version": "0.3.0-alpha.3",
```
Comment (piotr-oles) [inline packages/as-proto-gen/package.json]: please don't change that - we use `auto` for releases and it will automatically bump the version :)

File: packages/as-proto-gen/src/file-context.ts (hunk 2 above)
Comment (piotr-oles) [reply packages/as-proto-gen/src/file-context.ts:55]: Could we use built-in node API for this? https://nodejs.org/api/path.html#pathrelativefrom-to

File: packages/as-proto-gen/src/generate/enum.ts [hunk 14]
```diff
@@ -7,16 +7,17 @@ import * as assert from "assert";
...
   const enumValues = enumDescriptor.getValueList();
 
   return `
-    export enum ${Enum} {
+    export enum ${enumName} {
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/enum.ts]: why did you change that?

File: packages/as-proto-gen/src/generate/field.ts (hunk 4 above)
Comment (piotr-oles) [reply packages/as-proto-gen/src/generate/field.ts]: How adding `_` prevents name collision? 🤔 Could you provide an example of a name collision you had?

File: packages/as-proto-gen/src/generate/file.ts [hunk 15]
```diff
@@ -1,17 +1,24 @@
...
 import { FileContext } from "../file-context";
+import { GeneratorContext } from "../generator-context";
+import { getPathWithoutProto } from "../names";
 import { generateEnum } from "./enum";
-import * as assert from "assert";
+import { generateMessage } from "./message";
+import { version } from '../../package.json'
 
-export function generateFile(
+export function processFile(
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: why did you change the name of this function?

File: packages/as-proto-gen/src/generate/file.ts [hunk 16]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
-        }
-      `;
+  return `${fileContext.getImportsCode()}\n\n${types.join("\n\n")}`;
+}
+
+export function addFile(
+  filename: string,
+  code: string,
+  codeGenResponse: CodeGeneratorResponse,
+  protoc_version: string
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: nit: could you name it `protocVersion` to follow convention?

File: packages/as-proto-gen/src/generate/file.ts [hunk 17]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+  outputFile.setName(filename);
+  outputFile.setContent(formattedCode);
+  codeGenResponse.addFile(outputFile);
+}
+
+export function generateExport(
+  codeGenRequest: CodeGeneratorRequest,
+  codeGenResponse: CodeGeneratorResponse,
+  generatorContext: GeneratorContext,
+  protoc_version: string
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: nit: same here

File: packages/as-proto-gen/src/generate/file.ts [hunk 18]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+) {
+  const exports = new Map<string, Set<string>>();
+  const indexes = new Map<string, Set<string>>();
+
+  for (const filename of codeGenRequest.getFileToGenerateList()) {
+    const packages = generatorContext
+      .getFileDescriptorByFileName(filename)
+      ?.getPackage()
+      ?.split(".") as string[];
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: This assertion is dangerous given that file descriptor can be undefined. Could you change this line to `?.split(".") || []`?

File: packages/as-proto-gen/src/generate/file.ts [hunk 19]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+    const filePath = getPathWithoutProto(filename).split("/");
+
+    for (let i = 1; i < filePath.length; i++) {
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts:84]: could you add some comment what's going on here?

File: packages/as-proto-gen/src/generate/file.ts [hunk 20]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+      const path = filePath.slice(0, i).join("/");
+      const exportPath = filePath.at(i) as string;
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: `i` will always be positive, so maybe let's keep it simple and use `filePath[i]` :)

File: packages/as-proto-gen/src/generate/file.ts [hunk 21]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+      if (exports.has(path)) {
+        exports.get(path)?.add(exportPath);
+      } else {
+        exports.set(path, new Set<string>().add(exportPath));
+      }
+    }
+
+    for (let i = 0; i < filePath.length - 1; i++) {
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts:96]: same here - could you add some comment in the code what's going on in this loop?

File: packages/as-proto-gen/src/generate/file.ts [hunk 22]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+}
+
+export function generateExport(
+  codeGenRequest: CodeGeneratorRequest,
+  codeGenResponse: CodeGeneratorResponse,
+  generatorContext: GeneratorContext,
+  protoc_version: string
+) {
+  const exports = new Map<string, Set<string>>();
+  const indexes = new Map<string, Set<string>>();
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: IMO indexes is a little bit too generic name here - it looks like `filePathToPackages` or something in that vein. Or you could also use type aliases to document this, like:
```
type FilePath = string;
type PackageName = string;

// ...
  const indexes = new Map<FilePath, Set<PackageName>>();
```

File: packages/as-proto-gen/src/generate/file.ts [hunk 23]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+      }
+      if (indexes.has(path)) {
+        indexes.get(path)?.add(pkg);
+      } else {
+        indexes.set(path, new Set<string>().add(pkg));
+      }
+    }
+  }
+
+  for (const path of exports.keys()) {
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: could you add a comment in the code what's going on in this loop? :)

File: packages/as-proto-gen/src/generate/file.ts [hunk 24]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+    addFile(filename, code, codeGenResponse, protoc_version);
   }
 
-  return `
-    ${fileContext.getImportsCode()}
-    
-    ${NamespacedTypes}
-  `;
+  let topIndex = "";
+  indexes.forEach((pkgs: Set<String>, path: string) => {
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: nit: could you use `for (const [path, packages] of indexes) {` loop for consistency?

File: packages/as-proto-gen/src/generate/file.ts [hunk 25]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+    const filename = path + "/index.ts";
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: could you use built-in node.js API to manipulate paths? https://nodejs.org/api/path.html#pathjoinpaths

File: packages/as-proto-gen/src/generate/file.ts [hunk 26]
```diff
@@ -21,24 +28,111 @@ export function generateFile(
...
+    });
+    code += `export { ${[...pkgs].join(", ")} };`;
+    addFile(filename, code, codeGenResponse, protoc_version);
+
+    if (path.split("/").length == 1) {
+      topIndex += `export { ${[...pkgs].join(", ")} } from './${path}';\n`;
+    }
+  });
+
+  if (topIndex != "") {
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/file.ts]: what is topIndex?

File: packages/as-proto-gen/src/generate/message.ts (hunk 11 above)
Comment (piotr-oles) [reply packages/as-proto-gen/src/generate/message.ts:77]: to be fully correct, I think you should also add after `const Message = fileContext.registerDefinition(messageName, messageNamespace);` line:
```typescript
fileContext.registerDefinition(`decode${Message}`);
```
this ensures that there are no name conflicts in the file scope

File: packages/as-proto-gen/src/index.ts (hunk 6 above)
Comment (piotr-oles) [reply packages/as-proto-gen/src/index.ts:13]: maybe we could use something like `fs-extra` package - I think there are some improvements on fs error handling so maybe it would help?

File: packages/as-proto-gen/src/index.ts [hunk 27]
```diff
@@ -1,60 +1,95 @@
...
+    const generatorContext = new GeneratorContext();
 
-try {
-  const codeGenRequest = CodeGeneratorRequest.deserializeBinary(input);
-  const codeGenResponse = new CodeGeneratorResponse();
-  const generatorContext = new GeneratorContext();
+    const PROTOC_VERSION = codeGenRequest
+      .getCompilerVersion()
+      ?.toArray()
+      .slice(0, 3)
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/index.ts]: why slice?

File: packages/as-proto-gen/src/index.ts (hunk 7 above)
Comment (piotr-oles) [reply packages/as-proto-gen/src/index.ts]: Could you add documentation for this option to README.md?

**Review by piotr-oles — COMMENTED** [review]
I did a first pass of review :) Overall looks good, thanks for your work! I will do a second pass after you address these comments :)

File: packages/as-proto-gen/src/generate/field.ts (hunk 4 above)
Comment (Seb-figment) [reply packages/as-proto-gen/src/generate/field.ts]: This happen if a variable in a message is named the same as an import. This was a quick fix to just get going. I'll push a real fix.
[example](https://github.com/Seb-figment/as-proto-example/blob/8f0ed1b2705ac5135c8c31e6e6f50d6e56d566ce/src/bar/qux.ts#L30)

File: packages/as-proto-gen/src/generate/file.ts (hunk 15 above)
Comment (Seb-figment) [reply packages/as-proto-gen/src/generate/file.ts]: Just an artifact of when I was thinking of generating one file per proto package. I'll revert that.

File: packages/as-proto-gen/src/generate/file.ts (hunk 26 above)
Comment (Seb-figment) [reply packages/as-proto-gen/src/generate/file.ts]: topIndex is to generate the index file at root of package. Should have called the variable rootIndex ... will rename

File: packages/as-proto-gen/src/index.ts (hunk 27 above)
Comment (Seb-figment) [reply packages/as-proto-gen/src/index.ts]: not including the suffix return by getCompilerVersion. Most of the time this will be empty and didn't wanted version ending with a dot.  ie: `3.20.1.`

File: packages/as-proto-gen/src/generate/enum.ts (hunk 14 above)
Comment (Seb-figment) [reply packages/as-proto-gen/src/generate/enum.ts]: artifact. reverted.

**Comment by Seb-figment** [comment]
Pushed fixes for review comment. Please don't hesitate to correct the coding style. I've never used typescript before this.

**Comment by piotr-oles** [comment]
Thanks for the updates. I will review this next week :)

File: packages/as-proto-gen/src/generate/field.ts [hunk 28]
```diff
@@ -18,11 +18,11 @@ export function generateFieldEncodeInstruction(
 
   const fieldTag = getFieldTag(fieldDescriptor);
   const fieldName = generateFieldName(fieldDescriptor);
-  const fieldVariable = scopeContext.getFreeName(fieldName);
   const fieldTypeInstruction = generateFieldTypeInstruction(fieldDescriptor);
 
   if (isMessage) {
     const Message = generateRef(fieldDescriptor, scopeContext.getFileContext());
+    const fieldVariable = scopeContext.getFreeName(fieldName);
```
Comment (piotr-oles) [inline packages/as-proto-gen/src/generate/field.ts:25]: this seems to be unrelated - could you revert that?

**Review by piotr-oles — COMMENTED** [review]
I extracted some of unrelated changes to separate PRs and merged to main. This PR is still too complex for me to merge - could you split this to a few smaller ones? Maybe we could have separate one for proto dependencies, another for adding protoc version header, etc.?

File: packages/as-proto-gen/package.json [hunk 29]
```diff
@@ -21,8 +21,11 @@
     "as-proto-gen": "bin/as-proto-gen"
   },
   "devDependencies": {
+    "@types/fs-extra": "^9.0.13",
     "@types/google-protobuf": "^3.15.5",
-    "@types/prettier": "^2.4.2"
+    "@types/prettier": "^2.4.2",
+    "fs-extra": "^10.1.0",
+    "rimraf": "^3.0.2"
```
Comment (piotr-oles) [inline packages/as-proto-gen/package.json:28]: it's already in the workspace package.json - not needed here :)

File: packages/as-proto-gen/src/file-context.ts (hunk 2 above)
Comment (Seb-figment) [reply packages/as-proto-gen/src/file-context.ts:55]: We can't because the getRelativeImportPath also truncate path suffix to keep "namespacing" for imported object.

For example, if `google.protobuf.Timestamp` is imported from `cosmos/authz/v1beta1/authz.ts` it will be imported as `../../../google`

**Comment by Seb-figment** [comment]
@pienkowb will break this PR into smaller one.
